### PR TITLE
std(http/server): close open connections on server close

### LINKED ISSUE #3390: Issue with HTTP's close method
via @sholladay 

When I use serve() from http/server.ts, and I pass it port 0 to get a random available port, is there a clean way to determine which port was assigned?

Also, server.close() seems to wait for another request to come in before actually closing. If I call server.close(), Deno won't quit until I reload the page in the browser.

### LINKED ISSUE #3593: Server hangs on close
following code works at v0.27 but hangs at v0.28
```ts
import { test, runIfMain } from "https://deno.land/std@v0.27.0/testing/mod.ts";
import { assertEquals } from "https://deno.land/std@v0.27.0/testing/asserts.ts";
import { serve, Server } from "https://deno.land/std@v0.27.0/http/server.ts";

const encoder = new TextEncoder();
let s: Server = undefined;

async function createServer() {
  s = serve(":8080");
  for await (const req of s) {
    req.respond({ body: encoder.encode(req.url) });
  }
}

test(async function foo() {
  createServer();

  const res = await fetch("http://127.0.0.1:8080/hello");
  assertEquals(await res.text(), "/hello");
  s.close();
});

runIfMain(import.meta);
```

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes #3390

If you call `server.close()` on a server that has pending connections then Deno process won't exit until next request comes over for each connection.  

This PR ensures that all open connections that server owns get properly closed when server is closed.

### CODE REVIEW

File: cli/ops/net.rs [hunk 1]
```diff
@@ -179,7 +179,8 @@ fn op_shutdown(
   let shutdown_mode = match how {
     0 => Shutdown::Read,
     1 => Shutdown::Write,
-    _ => unimplemented!(),
+    2 => Shutdown::Both,
```
Comment (lucacasonato) [inline cli/ops/net.rs]: This has a different name in typescript. Would it make sense to unify rust to be `Shutdown::ReadWrite`?

File: std/http/server.ts [hunk 2]
```diff
@@ -375,12 +375,16 @@ export async function readRequest(
 
 export class Server implements AsyncIterable<ServerRequest> {
   private closing = false;
+  private connections: Conn[] = [];
```
Comment (ry) [inline std/http/server.ts]: We shouldn’t need to maintain a list of connections

File: std/http/server.ts [hunk 3]
```diff
@@ -375,12 +375,16 @@ export async function readRequest(
...
 
   constructor(public listener: Listener) {}
 
   close(): void {
     this.closing = true;
     this.listener.close();
+    for (const conn of this.connections) {
+      Deno.shutdown(conn.rid, Deno.ShutdownMode.ReadWrite);
+    }
```
Comment (ry) [inline std/http/server.ts]: Shutdown is inappropriate here

**Comment by nayeemrmn** [comment]
Is there some reason we shouldn't shutdown within the implementation of `conn.close()` as I suggested in https://github.com/denoland/deno/issues/3593#issuecomment-573413536? Calling `conn.close()` now makes all pending reads just hang, I don't see why that would ever be desired.

**Comment by bartlomieju** [comment]
Okay after some more thought I think problem described in #3593 is caused by two separate issues:
1) closing connection does not call `shutdown` on that socket causing all pending reads/writes to hang
2) `fetch` now uses shared HTTP client and **keeps connection open** for later reuse

File: std/http/server_test.ts [hunk 4]
```diff
@@ -640,6 +641,25 @@ test({
...
+      for await (const req of server) {
+        req.respond({ body: new TextEncoder().encode(req.url) });
+      }
+    }
+
+    const server = serve(":8123");
+    iteratorReq(server);
+    // fetch keeps open connection, this when calling `server.close()`
+    // server should close all open connections
+    const res = await fetch("http://127.0.0.1:8123/hello");
```
Comment (bartlomieju) [inline std/http/server_test.ts]: No longer true, rewrite it using `Deno.dial()` to present hanging of server on open connection

**Comment by nayeemrmn** [comment]
`server.close()` currently makes the server stop accepting new connections without closing existing ones. This patch makes it close existing connections as well. I don't think there's any more hanging to fix in doing so and the current behaviour is probably better, right?

**Comment by bartlomieju** [comment]
> `server.close()` currently makes the server stop accepting new connections without closing existing ones. This patch makes it close existing connections as well. I don't think there's any more hanging to fix in doing so and the current behaviour is probably better, right?

Let's see, just updated tests

**Comment by nayeemrmn** [comment]
@bartlomieju Okay I've seen it now (this behaviour:
> If you call `server.close()` on a server that has pending connections then Deno process won't exit until next request comes over for each connection.

). It won't stop trying to read requests on a connection until that connection closes... which is entirely in the client's hands. This still isn't necessarily a bad thing since people can implement their own timeouts etc. -- but at the same time there will a natural expectation that the iterators break on server close. Probably the right call to drop connections then 👍.

**Comment by nayeemrmn** [comment]
Just throwing this out there: an alternative solution would be to provide a separate `server.shutdown()` method to shutdown all connections and encourage people to use it after closing (maybe after a timeout) when they want to clean up completely. Safe simplicity vs versatility, I'm okay with either though.

**Comment by bartlomieju** [comment]
> Just throwing this out there: an alternative solution would be to provide a separate `server.shutdown()` method to shutdown all connections and encourage people to use it after closing (maybe after a timeout) when they want to clean up completely. Safe simplicity vs versatility, I'm okay with either though.

Yes, we discussed that offline, and indeed `server.shutdown()` is desirable method that would allow to stop accepting new connections and gracefully handle already active connections.

@nayeemrmn if you'd be interested in implementing that then using Go's approach is encouraged:
https://github.com/golang/go/blob/3743d2127040c283114a247da1319c3155a81f10/src/net/http/server.go#L2648-L2691

**Comment by bartlomieju** [comment]
Needs more investigation, closing for now

**Comment by bartlomieju** [comment]
With #4293 this PR can be moved forward

**Comment by bartlomieju** [comment]
@ry please review

File: std/http/server.ts [hunk 5]
```diff
@@ -124,12 +124,20 @@ export class ServerRequest {
...
   close(): void {
     this.closing = true;
     this.listener.close();
+    for (const conn of this.connections) {
+      try {
+        conn.close();
+      } catch (e) {
+        // Connection might have been already closed
+      }
+    }
```
Comment (ry) [inline std/http/server.ts]: Can you narrow the try-catch here?

```patch
         conn.close();
       } catch (e) {
         // Connection might have been already closed
+        if (!(e instanceof Deno.errors.BadResource)) {
+          throw e;
+        }
       }
     }
   }
```

Comment (ry) [reply std/http/server.ts]: actually, nevermind. it's fine.

**Pull request merged**